dask_cloudprovider/providers/gcp/instances.py

```diff
@@ -0,0 +1,326 @@
+
+    async def start(self):
+        await super().start()
+        await self.start_scheduler()
+
+    async def start_scheduler(self):
+        self.cluster._log(f"Launching cluster with the following configuration: " \
+                           f"\n  Source Image: {self.source_image} " \
+                           f"\n  Docker Image: {self.docker_image} " \
+                           f"\n  Machine Type: {self.machine_type} " \
+                           f"\n  Filesytsem Size: {self.filesystem_size} " \
+                           f"\n  N-GPU Type: {self.ngpus} {self.gpu_type}" \
+                           f"\n  Zone: {self.zone} " \
+                           )
+        self.cluster._log("Creating scheduler instance")
+        self.internal_ip, self.external_ip = await self.create_vm()
+        self.address = f"tcp://{self.external_ip}:8786"
+        await self.wait_for_scheduler()
+
+        # need to reserve internal IP for workers
+        # gcp docker containers can't see resolve ip address
+        self.cluster.scheduler_internal_ip = self.internal_ip
+        self.cluster.scheduler_external_ip = self.external_ip
+
+
+class GCPWorker(GCPInstance, WorkerMixin):
+    """Worker running in an GCP instance."""
+
+    def __init__(self, scheduler, *args, worker_command=None, worker_extra_args=None, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.worker_extra_args = self.config.get('worker_extra_args', [])
+        worker_command = ' '.join([worker_command] + self.worker_extra_args)
+        self.init_worker(scheduler, *args, worker_command=worker_command, **kwargs)
+
+    async def start(self):
+        await super().start()
+        await self.start_worker()
+
+    def init_worker(self, scheduler: str, *args, worker_command=None, **kwargs):
+        self.scheduler = scheduler
+        self.worker_command = worker_command
+
+        self.name = f"dask-worker-{str(uuid.uuid4())[:8]}"
+        self.command = (
+            f"{self.worker_command} {self.cluster.scheduler_internal_ip}:8786"
+        )
+
+        self.cluster._log(f"Starting worker: {self.name} with command: {self.command}")
+
+    async def start_worker(self):
+        self.cluster._log("Creating worker instance")
+        self.internal_ip, self.external_ip = await self.create_vm()
+        self.address = self.external_ip
+
+
+class GCPCluster(VMCluster):
+    """Cluster running on GCP Instances."""
+
+    def __init__(
+        self,
+        name="dask-gcp-example",
+        zone=None,
+        machine_type=None,
+        projectid=None,
+        source_image=None,
+        docker_image=None,
+        ngpus=None,
+        gpu_type=None,
+        worker_command="dask-cuda-worker",
+        worker_extra_args=None,
+        auto_shutdown=True,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        try:
+            self.compute = googleapiclient.discovery.build("compute", "v1")
```

Comment:
It seems like the only way to authenticate here is by setting the `GOOGLE_APPLICATION_CREDENTIALS` env var with a service account JSON.

It would be nice to be able to pass the path to a service account JSON as a kwarg (even if that calls `os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = path_to_creds`.

Ideally it would be nice to pick up the credentials created with `gcloud auth login` but not sure if that is possible.